sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetFileFormat.scala

```diff
@@ -105,29 +105,60 @@ object GpuParquetFileFormat {
     }
   }
 
-  def parquetWriterOptionsFromSchema[T <: NestedBuilder[_, _], V <: ParquetColumnWriterOptions]
-  (builder: ParquetColumnWriterOptions.NestedBuilder[T, V],
-   schema: StructType, writeInt96: Boolean): T = {
-    // TODO once https://github.com/rapidsai/cudf/issues/7654 is fixed go back to actually
-    // setting if the output is nullable or not everywhere we have hard-coded nullable=true
-    schema.foreach(field =>
-      field.dataType match {
+  def parquetWriterOptionsFromField[T <: NestedBuilder[_, _], V <: ParquetColumnWriterOptions](
+      builder: ParquetColumnWriterOptions.NestedBuilder[T, V],
+      dataType: DataType,
+      name: String,
+      writeInt96: Boolean,
+      nullable: Boolean): T = {
+      dataType match {
         case dt: DecimalType =>
-          builder.withDecimalColumn(field.name, dt.precision, true)
+          builder.withDecimalColumn(name, dt.precision, nullable)
         case TimestampType =>
-          builder.withTimestampColumn(field.name,
-            writeInt96, true)
+          builder.withTimestampColumn(name, writeInt96, nullable)
         case s: StructType =>
           builder.withStructColumn(
-            parquetWriterOptionsFromSchema(structBuilder(field.name), s, writeInt96).build())
+            parquetWriterOptionsFromSchema(
+              structBuilder(name, nullable),
+              s,
+              writeInt96).build())
         case a: ArrayType =>
           builder.withListColumn(
-            parquetWriterOptionsFromSchema(listBuilder(field.name),
-              StructType(Array(StructField(field.name, a.elementType, true))), writeInt96)
-              .build())
+            parquetWriterOptionsFromField(
+              listBuilder(name, nullable),
+              a.elementType,
+              name,
+              writeInt96,
+              true).build())
+        case m: MapType =>
+          builder.withMapColumn(
+            mapColumn(name,
+              parquetWriterOptionsFromField(
+                ParquetWriterOptions.builder(),
+                m.keyType,
+                "key",
+                writeInt96,
+                false).build().getChildColumnOptions()(0),
```

Comment:
I am a little concerned with this in relation to https://github.com/rapidsai/cudf/issues/7654. The problem is even if the key column contains no nulls, if it has a validity buffer, that is enough to cause problems, and I don't know enough to guarantee that it will not have one.  Have we tested to see if we set this to true if Spark has issues with reading it?

As an alternative we could take the hit ourselves and write a cudf function that will walk through an input column view and prune off unneeded validity buffers.  That would let us call that before writing to parquet/etc. And would not add any extra overhead to the core cudf.

Reply:
It's working fine after I set the key column to be nullable. 

I will take on the alternative as a follow-on if I can get to it this sprint because I know its come up many times.

Reply:
I apologize, I had put the nullable in the wrong place, its confusing when we have multiple nesting. Upon testing it again, spark doesn't like it and is throwing an exception like @revans2 suspected. 
```
E                     File "/home/rjafri/spark-3.1.1-bin-hadoop3.2/python/lib/pyspark.zip/pyspark/sql/types.py", line 1367, in verify_map
E                     File "/home/rjafri/spark-3.1.1-bin-hadoop3.2/python/lib/pyspark.zip/pyspark/sql/types.py", line 1408, in verify
E                     File "/home/rjafri/spark-3.1.1-bin-hadoop3.2/python/lib/pyspark.zip/pyspark/sql/types.py", line 1278, in verify_nullability
E                   ValueError: key of map field _c57: This field is not nullable, but got None

```
Are you OK with getting this in while I work on the cudf function?

Reply:
No I am not.

Reply:
OK, I am working on a cudf PR. The changes will be in the TableJni.cpp, we can later move it to the core cudf-cpp if there is a need. Is that OK?